repose-aggregator/docs/build.gradle

```diff
@@ -0,0 +1,286 @@
+    from("${project.rootDir}/repose-aggregator/artifacts/docker/src/docker/centos/Dockerfile")
+
+    // Place the file into a resources subdirectory.
+    into("${project.buildDir}/asciibinder/_package/main/${version}/resources/centos/")
+}
+
+task asciibindersearch(dependsOn: [copyXsds, copyCentosDockerfile, copyUbuntuDockerfile], group: 'build', description: 'Builds the search for the latest version of the docs') {
+    inputs.dir(file("${project.buildDir}/asciibinder/_package/main/${version}"))
+    outputs.files(
+        file("${project.buildDir}/asciibinder/_package/main/search.html"),
+        file("${project.buildDir}/asciibinder/_package/main/data_${version}.json"),
+        file("${project.buildDir}/asciibinder/_package/main/versions.json")
+    )
+    doLast {
+        def buildProcess =
+            "docker build --tag local/asciibindersearch ${project.projectDir}/src/docker/Python3/"
+                .execute(null, file("${project.buildDir}/asciibinder"))
+        buildProcess.waitFor()
+        if (buildProcess.exitValue() != 0) {
+            logger.error new String(buildProcess.getInputStream().bytes)
+            logger.error new String(buildProcess.getErrorStream().bytes)
+        } else {
+            logger.info new String(buildProcess.getInputStream().bytes)
+        }
+
+        def runProcess =
+            "docker run --tty --user ${uid}:${gid} --volume ${project.buildDir}/asciibinder:/docs:z local/asciibindersearch"
+                .execute(null, file("${project.buildDir}/asciibinder"))
+        runProcess.waitFor()
+        if (runProcess.exitValue() != 0) {
+            logger.error new String(runProcess.getInputStream().bytes)
+            logger.error new String(runProcess.getErrorStream().bytes)
+        } else {
+            logger.info new String(runProcess.getInputStream().bytes)
+        }
+    }
+}
+
+task stageDocs(dependsOn: asciibindersearch) {
+    inputs.dir(file("${project.projectDir}/src/resources"))
+    inputs.dir(file("${project.buildDir}/asciibinder/_package/main"))
+    outputs.dir(file("${project.buildDir}/gh-pages"))
+    doLast {
+        def workingDir = file("${project.buildDir}/gh-pages")
+
+        // IF the gh-pages dir exists,
+        // THEN just clean it up;
+        // ELSE clone the repo and checkout the correct branch.
+        if (file("${project.buildDir}/gh-pages/.git").exists()) {
+            workingDir.listFiles(new FilenameFilter() {
+                @Override
+                boolean accept(File dir, String name) {
+                    return !(name == '.git')
+                }
+            }).each { file ->
+                if (file.isDirectory()) {
+                    file.deleteDir()
+                } else {
+                    file.delete()
+                }
+            }
+            def grgit = Grgit.open(dir: workingDir.getAbsolutePath())
+            grgit.checkout(branch: 'gh-pages')
+            grgit.reset(mode: 'HARD')
```

Comment:
Since the `force` options are not present in the `Grgit` operations, I updated this to the equivalent of:
```
git reset --hard
git checkout gh-pages
git pull
git clean -d
```

Reply:
Unfortunately, the new flow is not quite fool-proof. Consider the case where the `gh-pages` branch contains a file added in some commit. The revision did not contain that file in some commit, but that file was added and is currently hanging out untracked with contents that differ from the file on the `gh-pages` branch. `reset --hard` will not remove the file since it is untracked, but when we go to checkout `gh-pages`, we fail because the changes would be overwritten. That's taken care of when using `checkout -f`, but not `reset --hard`. If we can't use `checkout -f`, then we would likely have to use both `reset --hard` and `clean -d` before checking out another branch.

Git workflows, yo.

Reply:
Updated.